hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java

```diff
@@ -0,0 +1,852 @@
+package com.hazelcast.config;
+
+import com.hazelcast.config.CacheSimpleConfig.ExpiryPolicyFactoryConfig;
+import com.hazelcast.config.CacheSimpleConfig.ExpiryPolicyFactoryConfig.DurationConfig;
+import com.hazelcast.config.CacheSimpleConfig.ExpiryPolicyFactoryConfig.TimedExpiryPolicyFactoryConfig;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.util.CollectionUtil;
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static java.text.MessageFormat.format;
+
+class ConfigCompatibilityChecker {
+
+    /**
+     * Checks if two {@link Config}'s are compatible. This mostly means that the config values will have the same
+     * impact on the behaviour of the system but are not necessarily the same (e.g. null value is sometimes the same
+     * as an empty collection or a disabled config).
+     * NOTE : This method checks MOST but NOT ALL configuration. As such it is best used in test scenarios to cover
+     * as much config checks as possible automatically.
+     *
+     * @param c1 the {@link Config} to check
+     * @param c2 the {@link Config} to check
+     * @return {@code true} if the configs are compatible
+     * @throws HazelcastException       if configs are incompatible
+     * @throws IllegalArgumentException if one of the configs is {@code null}
+     */
+    static boolean isCompatible(final Config c1, final Config c2) {
+        if (c1 == c2) {
+            return true;
+        }
+        if (c1 == null || c2 == null) {
+            throw new IllegalArgumentException("One of the two configs is null");
+        }
+        if (!c1.getGroupConfig().getName().equals(c2.getGroupConfig().getName())) {
+            return false;
+        }
+        if (!c1.getGroupConfig().getPassword().equals(c2.getGroupConfig().getPassword())) {
+            throw new HazelcastException("Incompatible group password");
+        }
+        checkPartitionGroupConfigCompatible(c1, c2);
+        checkSerializationConfigCompatible(c1, c2);
+        checkServicesConfigCompatible(c1, c2);
+        checkManagementCenterConfigCompatible(c1, c2);
+        checkHotRestartConfigCompatible(c1, c2);
+
+        checkNetworkConfigCompatible(c1, c2);
+        checkCompatibleConfigs("map", c1, c2, c1.getMapConfigs(), c2.getMapConfigs(), new MapConfigChecker());
```

Comment:
Nice & arduous job. Would it be possible to homogenize all config checks to be invoked with `checkCompatibleConfigs`? (as far as I could tell, all compatibility checks are done by `ConfigChecker`s)